packages/dds/tree/docs/wip/revertible-fork-api.md

```diff
@@ -0,0 +1,207 @@
+```
+
+## Garbage Collection
+
+However, there is an important reason that the API _can't_ be so simple, and that reason is the root cause of much of the complexity that has emerged.
+It is potentially _expensive_, in both memory and network bandwidth, to retain the information that is required to revert an arbitrary commit.
+
+For example, suppose a commit deletes a massive amount of content from the document.
+That content cannot _actually_ be deleted from the user's program memory.
+If the commit were to be reverted, then that massive amount of content would need to be re-inserted, and thus needs to be retained.
+
+Additionally, consider that the "history" (i.e. the sequence of commits that have applied to a document so far) will grow unboundedly so long as every commit is eligible to be reverted. Those commits can never be dropped from memory, because the user might want to revert a commit from a "very long time ago".
+
+### Revertibles
+
+This is solved by exposing handles to commits which can (and should) be disposed when they are no longer possibly going to be reverted.
+Since these handles are specifically used by the revert functionality, we call them `Revertibles`, and they have (something like) the following interface:
+
+```ts
+interface Revertible {
+	revert();
+	dispose();
+}
+```
+
+Typically, revertibles are placed in an ordered undo and redo stack and managed by the user (perhaps via an undo/redo helper library).
+
+> Revertibles need not be reverted in order.
+> An undo/redo stack is a common use of revertibles, but other "undo-ish" data structures might be appropriate for an application as well. 
+
+Importantly, revertibles are only allocated if the user requests them to be.
+When a new commit is applied to the tree, an event (`"commitApplied"`) fires which provides the user a _function_ that they can call _if_ they want to allocate a revertible.
+If they do this, then they are also responsible for disposing the revertible as soon as they no longer need it (for example, if it ends up too far back in their undo stack).
+In this way, revertibles are given the minimal lifetime possible (if they are given a life at all) in order to reduce the overhead of maintaining them.
+
+## Branches
+
+How do revertibles interact with SharedTree [branches](../main/indexes-and-branches.md)?
+Note that a `Revertible` has a parameterless `revert()` method.
+Under the hood it is performing the fundamental `revert(commit, branch)` operation and therefore the revertible must be implicitly associated with both a commit _and_ a branch. This means that different branches will have different revertible objects, even for reverting the same commit.
+
+This is accomplished by allowing revertibles to be _cloned_ (or "forked"). Cloning a revertible produces a new revertible which is associated with the provided branch - so, when `revert()` is called on that new revertible, it applies the reverted commit to the head of that branch.
```

Comment:
It's not clear from the text here that this cloning process involves providing another branch for the clone to be associated with.

Reply:
I guess it's covered in detail below, but it would still be nice to clarify here.